### PR TITLE
Sync ``libasr`` with LPython

### PR DESCRIPTION
**Pull request by czgdp1807**
Corresponding LPython PR - https://github.com/lcompilers/lpython/pull/1411

### CODE REVIEW

**Comment by czgdp1807** [comment]
@Smit-create @Shaikh-Ubaid @Thirumalai-Shaktivel Please review the changes and let me know if I something needs to be reverted or fixed. Thanks.

File: src/libasr/pass/nested_vars.cpp [hunk 1]
```diff
@@ -238,29 +238,19 @@ class NestedVarVisitor : public ASR::BaseWalkVisitor<NestedVarVisitor>
                 calls_to.end()){
                 calls_to.push_back(call_hash);
             }
-            for (size_t i=0; i<x.n_args; i++) {
-                if (ASR::is_a<ASR::Var_t>(*x.m_args[i].m_value)) {
-                    visit_Var(*ASR::down_cast<ASR::Var_t>(x.m_args[i].m_value));
-                }
-            }
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/pass/nested_vars.cpp:245]: This and SubroutineCall changes are to be reverted, I think.

File: src/libasr/pass/nested_vars.cpp [hunk 2]
```diff
@@ -228,7 +228,7 @@ class NestedVarVisitor : public ASR::BaseWalkVisitor<NestedVarVisitor>
     }
 
     void visit_FunctionCall(const ASR::FunctionCall_t &x) {
-        if (nesting_depth) {
+        if (nesting_depth == 1){
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/pass/nested_vars.cpp:231]: ..

File: src/libasr/codegen/asr_to_c_cpp.h [hunk 3]
```diff
@@ -1487,77 +1580,9 @@ R"(#include <stdio.h>
...
-    {
-        std::string indent(indentation_level * indentation_spaces, ' ');
-        this->visit_expr(*x.m_test);
-        std::string var = std::move(src);
-        std::string out = indent + "if (";
-
-        for (size_t i = 0; i < x.n_body; i++) {
-            if (i > 0)
-                out += indent + "else if (";
-            ASR::case_stmt_t* stmt = x.m_body[i];
```
Comment (Smit-create) [inline src/libasr/codegen/asr_to_c_cpp.h:1500]: `visit_Select` needs a fix here.

File: src/libasr/pass/pass_manager.h [hunk 4]
```diff
@@ -126,7 +128,7 @@ namespace LCompilers {
                 "forall",
                 "select_case",
                 "inline_function_calls",
-                "unused_functions"
+                "unused_functions",
```
Comment (Smit-create) [inline src/libasr/pass/pass_manager.h:131]:
```suggestion
                "unused_functions"
```

File: src/libasr/asdl_cpp.py [hunk 5]
```diff
@@ -1399,42 +1394,41 @@ def visitField(self, field, cons):
...
+                    self.emit('            dec_indent();', level+1)
+                    self.emit('            s.append("\\n" + indtd);', level+1)
                     self.emit('        }', level)
                     self.emit('        i++;', level)
                     self.emit('    }', level)
                     self.emit('}', level)
                     self.emit('if(indent) {', level)
                     self.emit(    'dec_indent();', level+1)
-                    self.emit(    's.append("\\n" + indented);', level+1)
+                    self.emit(    's.append("\\n" + indtd);', level+1)
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/asdl_cpp.py:1431]: The changes in asdl_cpp.py are to be reverted.

**Comment by czgdp1807** [comment]
@lfortran/libasrsyncreviewers @akshanshbhatt The tests pass on this PR. Please see the updated diff (commit by commit) and let me know if there is something that needs to be reverted or fixed. Thanks.

File: tests/reference/cpp-expr_05-3b5ec63.stdout [hunk 6]
```diff
@@ -72,7 +72,7 @@ void main2() {
     x = std::pow(a + b, 2);
     x = (a + b)*(std::pow(a, 2) - a*b + std::pow(b, 2));
     x = (a - b)*(a + b)*(std::pow(a, 2) + std::pow(b, 2));
-    x = 1/(a*b);
+    x = 1/a*b;
```
Comment (akshanshbhatt) [inline tests/reference/cpp-expr_05-3b5ec63.stdout:75]: Is the updated behavior intended? Both will generate different results imo.

File: tests/reference/c-expr_11-8e5ae80.stdout [hunk 7]
```diff
@@ -38,6 +38,6 @@ struct dimension_descriptor
 int main(int argc, char* argv[])
 {
     double x;
-    x = (  2.00000000000000000e+00*x +   1.00000000000000000e+00)/(x*(x + (double)(1)));
+    x = (  2.00000000000000000e+00*x +   1.00000000000000000e+00)/x*(x + (double)(1));
```
Comment (akshanshbhatt) [inline tests/reference/c-expr_11-8e5ae80.stdout:41]: This has a similar problem as the last one. It will generate different results.

**Comment by certik** [comment]
Awesome, great job!!

**Pull request merged**